include/dali/core/fast_div.h

```diff
@@ -0,0 +1,247 @@
+  if (sh == 0) {
+    return x;
+  } else if (sh >= bits) {
+    return { 0, x.lo << (sh-bits) };
+  } else {
+    return {
+      x.lo << sh,
+      x.hi << sh | U(x.lo) >> (bits - sh)
+    };
+  }
+}
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> operator>>(lohi<T> x, unsigned sh) {
+  using U = typename std::make_unsigned<T>::type;
+  static constexpr unsigned bits = sizeof(T) * 8;
+  if (sh == 0) {
+    return x;
+  } else if (sh >= bits) {
+    return { x.hi >> (sh-bits), 0 };
+  } else {
+    return {
+      U(x.lo) >> sh | x.hi << (bits - sh),
+      x.hi >> sh
+    };
+  }
+}
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> &operator<<=(lohi<T> &x, unsigned sh) {
+  x = x << sh;
+  return x;
+}
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> &operator>>=(lohi<T> &x, unsigned sh) {
+  x = x >> sh;
+  return x;
+}
+
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE lohi<T> operator-(lohi<T> a, lohi<T> b) {
+  lohi<T> ret;
+  ret.lo = a.lo - b.lo;
+  int borrow = (b.lo && ret.lo > a.lo);
+  ret.hi = a.hi - b.hi - borrow;
+  return ret;
+}
+
+DALI_HOST_DEV inline uint32_t div_lohi(uint32_t lo, uint32_t hi, uint32_t divisor) {
+  return (static_cast<uint64_t>(hi) << 32 | lo) / divisor;
+}
+
+DALI_HOST_DEV DALI_FORCEINLINE lohi<uint64_t> mull(uint64_t a, uint64_t b) {
+  lohi<uint64_t> ret;
+#ifdef __CUDA_ARCH__
+  ret.lo = a * b;
+  ret.hi = __umul64hi(a, b);
+#else
+  unsigned __int128 m = (unsigned __int128)a * b;
+  ret.lo = m;
+  ret.hi = m >> 64;
+#endif
+  return ret;
+}
+
+using dali::ilog2;
+
+template <typename uint>
+DALI_HOST_DEV DALI_FORCEINLINE int ilog2(lohi<uint> lh) {
+  return lh.hi ? ilog2(lh.hi) + sizeof(uint)*8 : ilog2(lh.lo);
+}
+
+DALI_HOST_DEV inline uint64_t div_lohi(uint64_t lo, uint64_t hi, uint64_t divisor) {
+#if defined(__x86_64) && defined(__GNUC__) && !defined(__CUDA_ARCH__)
+  // I hope this gets compiled to dividing rdx:rax register pair by a 64-bit value
+  return (static_cast<unsigned __int128>(hi) << 64 | lo) / divisor;
+#else
+  // NVCC doesn't support __int128 in device code, so we need a bit of hackery
+  if (!hi)  // No high part? Just divide in 64-bits and be done.
+    return lo / divisor;
+
+  // long division:
+
+  int lnum = ilog2(hi) + 64;
+  int lden = ilog2(divisor);
+
+  lohi<uint64_t> num = { lo, hi };
+  lohi<uint64_t> den = { divisor, 0 };
+
+  // calculate MSB positions...
+  int sh = lnum - lden;
+  // .. and align numerator and denominator
+  den <<= sh;
+
+  uint64_t q = 0;
+
+  while (sh >= 0) {
+    lohi<uint64_t> dif = num - den;  // this serves both as difference and comparison
+    if (static_cast<int64_t>(dif.hi) >= 0) {
+      num = dif;
+      q |= static_cast<uint64_t>(1) << sh;
+    }
+    sh--;
+    den >>= 1;
+  }
+  return q;
+#endif
+}
+
+}  // namespace detail
+
+/**
+ * @brief Fast unsigned integer division
+ *
+ * Based on:
+ * Labor of Division (Episode III): Faster Unsigned Division by Constants
```

Comment:
I don't know if there is value in providing a direct link too (it is easier than googling but may get outdated soon).

Reply:
Done.